rust/lance/src/io/commit.rs

```diff
@@ -246,28 +247,34 @@ pub(crate) async fn migrate_fragments(
 ///
 /// Indices might be missing `fragment_bitmap`, so this function will add it.
 async fn migrate_indices(dataset: &Dataset, indices: &mut [Index]) -> Result<()> {
-    // Early return so we have a fast path if they are already migrated.
-    if indices.iter().all(|i| i.fragment_bitmap.is_some()) {
-        return Ok(());
-    }
-
     for index in indices {
-        if index.fragment_bitmap.is_none() {
-            // Load the read version of the index
-            let old_version = match dataset.checkout_version(index.dataset_version).await {
-                Ok(dataset) => dataset,
-                // If the version doesn't exist anymore, skip it.
-                Err(crate::Error::DatasetNotFound { .. }) => continue,
-                // Any other error we return.
-                Err(e) => return Err(e),
+        let mut must_recalculate_fragment_bitmap = index.fragment_bitmap.is_none();
+        if !must_recalculate_fragment_bitmap {
+            match dataset.checkout_version(index.dataset_version).await {
+                Ok(old_dataset) => {
+                    must_recalculate_fragment_bitmap |=
+                        if let Some(writer_version) = &old_dataset.manifest.writer_version {
+                            writer_version.older_than(0, 8, 15)
+                        } else {
+                            true
+                        };
+                }
+                Err(crate::Error::NotFound { .. }) | Err(crate::Error::DatasetNotFound { .. }) => {
+                    must_recalculate_fragment_bitmap = true;
+                }
```

Comment:
I'm worried about this code path. Datasets that regularly run `cleanup_old_versions()` and `optimize_indices()` will hit this a lot, won't they? Maybe we should actually attach a writer version to the indices themselves? That way we can tell whether we can trust the index even if the version it was originally created in is long gone. Indexes might be quite long-lived, right?

Reply:
> Datasets that regularly run cleanup_old_versions() and optimize_indices() will hit this a lot, won't they?

Yes

> Maybe we should actually attach a writer version to the indices themselves?

Let me give that a try real quick

> Indexes might be quite long-lived, right?

Yes (not any single UUID but the conceptual index yes)